### PR TITLE
Implements Parallel Event Processing for LiveClient and Some Examples

### PR DESCRIPTION
**Pull request by dvonthenen**
Implements: https://github.com/deepgram/deepgram-dotnet-sdk/issues/198, https://github.com/deepgram/deepgram-dotnet-sdk/issues/195, https://github.com/deepgram/deepgram-dotnet-sdk/issues/193

This implements the following:
- Parallel Event Processing for multiple event subscribers (just implements a function and uses it in Live/Client.cs)
- More of this PR are examples for:
  - PreRecorded (SITS)
  - Analyze (SITS)
  - Speak
  - Edge Cases (ie Prove out KeepAlives work and timeouts will be triggered)
- Updated the logging format to something more parsable

## Summary by CodeRabbit

- **New Features**
	- Expanded project scope to include new testing, edge cases, and analysis features such as sentiment, intent, summary, and topic analysis.
	- Introduced real-time text-to-speech API "Deepgram Aura" for conversational voice AI agents.
	- Added functionality for analyzing audio data for intents, sentiment, summaries, and topics using the Deepgram .NET SDK.
	- Implemented functionality for prerecorded audio transcription and analysis across various dimensions (intent, sentiment, summary, topic).
	- Added a new `Speak` project with a `woodchuck` subproject for text-to-speech processing.
- **Improvements**
	- Refactored client methods for improved event handling and parallel processing.
	- Updated logging configuration for more detailed output.
- **Bug Fixes**
	- Adjusted methods for handling cancellation tokens and disposing of WebSocket connections correctly.
- **Documentation**
	- Enhanced examples and project files to better demonstrate the usage of the Deepgram .NET SDK for audio analysis and transcription.

### CODE REVIEW

File: Deepgram.Dev.sln [hunk 1]
```diff
@@ -35,6 +33,70 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "url", "url", "{0414D1CF-79F
...
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "PreRecorded", "examples\prerecorded\sentiment\PreRecorded.csproj", "{ABBF01EA-1015-4C16-A68C-F30A3904799B}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "PreRecorded", "examples\prerecorded\summary\PreRecorded.csproj", "{27C58ED4-DB4E-4A38-8D0E-D14854FE1814}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "PreRecorded", "examples\prerecorded\topic\PreRecorded.csproj", "{C80145E0-5E37-4FA6-83FD-C9264876CAAF}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "PreRecorded", "examples\prerecorded\url\PreRecorded.csproj", "{1E5D1117-C170-4E6A-BC41-F7F8F46F2C59}"
+EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "ana;lyze", "ana;lyze", "{83D6D1F7-97E1-4E35-A0AA-E3ED5977EA72}"
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply Deepgram.Dev.sln:72]: Will fix!

File: Deepgram/Clients/Live/v1/Client.cs [hunk 2]
```diff
@@ -715,15 +703,15 @@
...
+    internal void InvokeParallel<T>(EventHandler<T> eventHandler, T e)
+    {
+        if (eventHandler != null)
+        {
+            Parallel.ForEach(
+                eventHandler.GetInvocationList().Cast<EventHandler<T>>(),
+                (handler) =>
+                    handler(null, e));
+        }
+    }
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply Deepgram/Clients/Live/v1/Client.cs:715]: good suggestion!

File: examples/prerecorded/sentiment/Program.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        static async Task Main(string[] args)
+        {
+            // Initialize Library with default logging
+            // Normal logging is "Info" level
+            Library.Initialize();
+
+            // Set "DEEPGRAM_API_KEY" environment variable to your Deepgram API Key
+            var deepgramClient = new PreRecordedClient();
+
+            var audioData = File.ReadAllBytes(@"CallCenterPhoneCall.mp3");
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/prerecorded/sentiment/Program.cs:22]: Will update!

File: tests/edge_cases/reconnect_same_object/ReconnectStreaming.csporj [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    <Nullable>enable</Nullable>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\..\..\Deepgram\Deepgram.csproj" />
+  </ItemGroup>
+  
+  <ItemGroup>
+    <Using Include="Deepgram" />
+  </ItemGroup>
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply tests/edge_cases/reconnect_same_object/ReconnectStreaming.csporj:16]: Will fix!

File: tests/edge_cases/reconnect_same_object/ReconnectStreaming.csporj [hunk 5]
```diff
@@ -0,0 +1,18 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply tests/edge_cases/reconnect_same_object/ReconnectStreaming.csporj:1]: Will update!

File: examples/prerecorded/intent/PreRecorded.csproj [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+	<PropertyGroup>
+		<OutputType>Exe</OutputType>
+		<TargetFramework>net6.0</TargetFramework>
+		<ImplicitUsings>enable</ImplicitUsings>
+		<Nullable>enable</Nullable>	
+	</PropertyGroup>
+	
+	<ItemGroup>	
+		<PackageReference Include="Microsoft.Extensions.Http" Version="8.0.0" />
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/prerecorded/intent/PreRecorded.csproj:11]: After I tried this, it started to give me compile errors. Ignoring.

File: tests/expected_failures/exercise_timeout/Program.cs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+namespace SampleApp
+{
+    class Program
+    {
+        static async Task Main(string[] args)
+        {
+            // Normal logging is "Info" level
+            Library.Initialize();
+            // OR very chatty logging
+            //Library.Initialize(LogLevel.Debug); // LogLevel.Default, LogLevel.Debug, LogLevel.Verbose
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply tests/expected_failures/exercise_timeout/Program.cs:17]: This is just reference code intentionally left in there.

File: tests/expected_failures/exercise_timeout/Program.cs [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+                SampleRate = 16000,
+                Punctuate = true,
+                SmartFormat = true,
+            };
+            await liveClient.Connect(liveSchema);
+
+            // Wait for 20 seconds
+            Console.WriteLine("\n\nWe are intentionally looking to timeout here to exercise the functionality.");
+            Console.WriteLine("This should take less than 20 seconds to timeout.");
+            Thread.Sleep(20000);
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply tests/expected_failures/exercise_timeout/Program.cs:60]: Will fix!

File: examples/prerecorded/sentiment/Program.cs [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+{
+    class Program
+    {
+        static async Task Main(string[] args)
+        {
+            // Initialize Library with default logging
+            // Normal logging is "Info" level
+            Library.Initialize();
+
+            // Set "DEEPGRAM_API_KEY" environment variable to your Deepgram API Key
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/prerecorded/sentiment/Program.cs:19]: Ignoring.

File: examples/speak/file/woodchuck/Program.cs [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+{
+    class Program
+    {
+        static async Task Main(string[] args)
+        {
+            // Initialize Library with default logging
+            // Normal logging is "Info" level
+            Library.Initialize();
+
+            // Set "DEEPGRAM_API_KEY" environment variable to your Deepgram API Key
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/speak/file/woodchuck/Program.cs:20]: Ignoring.

File: Deepgram.Dev.sln [hunk 11]
```diff
@@ -35,6 +33,70 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "url", "url", "{0414D1CF-79F
...
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "woodchuck", "woodchuck", "{1826908E-F3F3-4146-A2A1-5B3D71CB7F3D}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Speak", "examples\speak\file\woodchuck\Speak.csproj", "{BE44FCAC-FBFF-4D70-BE52-B181049D1F70}"
+EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "manage", "manage", "{FA5723B3-74E9-4221-80EF-4833C1C3DD9F}"
+EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "balances", "balances", "{7CD5D816-0F30-45D7-9C9F-872E67F6A711}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Manage", "examples\manage\balances\Manage.csproj", "{2BF14C9B-755E-41FD-BB59-D6A82E0EFC51}"
+EndProject
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply Deepgram.Dev.sln:99]: Ignoring. These are the folder within the VS project file.

File: examples/prerecorded/url/Program.cs [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+            // Set "DEEPGRAM_API_KEY" environment variable to your Deepgram API Key
+            var deepgramClient = new PreRecordedClient();
+
+            var response = await deepgramClient.TranscribeUrl(
+                new UrlSource("https://static.deepgram.com/examples/Bueller-Life-moves-pretty-fast.wav"),
+                new PrerecordedSchema()
+                {
+                    Model = "nova-2",
+                });
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/prerecorded/url/Program.cs:27]: This is just a simple example. Ignoring.

File: examples/analyze/sentiment/Program.cs [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+        static async Task Main(string[] args)
+        {
+            // Initialize Library with default logging
+            // Normal logging is "Info" level
+            Library.Initialize();
+
+            // Set "DEEPGRAM_API_KEY" environment variable to your Deepgram API Key
+            var deepgramClient = new AnalyzeClient();
+
+            var audioData = File.ReadAllBytes(@"conversation.txt");
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/sentiment/Program.cs:29]: this was already updated

File: tests/edge_cases/reconnect_same_object/Program.cs [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+    class Program
+    {
+        static async Task Main(string[] args)
+        {
+            // Initialize Library with default logging
+            // Normal logging is "Info" level
+            //Deepgram.Library.Initialize();
+            // OR very chatty logging
+            Deepgram.Library.Initialize(LogLevel.Debug); // LogLevel.Default, LogLevel.Debug, LogLevel.Verbose
+            Deepgram.Microphone.Library.Initialize();
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply tests/edge_cases/reconnect_same_object/Program.cs:24]: I need to revisit this example anyways. Ignoring for now.

File: examples/analyze/summary/Program.cs [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+        static async Task Main(string[] args)
+        {
+            // Initialize Library with default logging
+            // Normal logging is "Info" level
+            Library.Initialize();
+
+            // Set "DEEPGRAM_API_KEY" environment variable to your Deepgram API Key
+            var deepgramClient = new AnalyzeClient();
+
+            var audioData = File.ReadAllBytes(@"conversation.txt");
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/summary/Program.cs:29]: This is already included.

File: examples/analyze/summary/Program.cs [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+{
+    class Program
+    {
+        static async Task Main(string[] args)
+        {
+            // Initialize Library with default logging
+            // Normal logging is "Info" level
+            Library.Initialize();
+
+            // Set "DEEPGRAM_API_KEY" environment variable to your Deepgram API Key
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/summary/Program.cs:19]: Ignoring.

File: examples/analyze/intent/conversation.txt [hunk 17]
```diff
@@ -0,0 +1,71 @@
+Meet Deepgram Aura: real-time text-to-speech for real-time AI agents
+----------
+It’s been a year since large language models (LLMs) seemingly went mainstream overnight (Happy Birthday, ChatGPT!!!), and the world has witnessed both rapid development of these technologies and immense interest in their potential. We believe that we have reached an inflection point where voice-based interfaces will be the primary means to accessing LLMs and the experiences they unlock. Here are a few recent signals in support of our thesis:
+
+- Good old fashioned voice notes are enjoying a healthy resurgence.
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/intent/conversation.txt:5]: this is just sample text to analyze

File: examples/prerecorded/topic/Program.cs [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+{
+    class Program
+    {
+        static async Task Main(string[] args)
+        {
+            // Initialize Library with default logging
+            // Normal logging is "Info" level
+            Library.Initialize();
+
+            // Set "DEEPGRAM_API_KEY" environment variable to your Deepgram API Key
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/prerecorded/topic/Program.cs:19]: ignoring.

File: examples/prerecorded/intent/Program.cs [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+        static async Task Main(string[] args)
+        {
+            // Initialize Library with default logging
+            // Normal logging is "Info" level
+            Library.Initialize();
+
+            // Set "DEEPGRAM_API_KEY" environment variable to your Deepgram API Key
+            var deepgramClient = new PreRecordedClient();
+
+            var audioData = File.ReadAllBytes(@"CallCenterPhoneCall.mp3");
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/prerecorded/intent/Program.cs:29]: This already exists.

File: examples/analyze/topic/conversation.txt [hunk 20]
```diff
@@ -0,0 +1,71 @@
+Meet Deepgram Aura: real-time text-to-speech for real-time AI agents
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/topic/conversation.txt:1]: this is just sample text to analyze

File: examples/analyze/summary/conversation.txt [hunk 21]
```diff
@@ -0,0 +1,71 @@
+Meet Deepgram Aura: real-time text-to-speech for real-time AI agents
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/summary/conversation.txt:1]: this is just sample text to analyze

File: examples/analyze/sentiment/conversation.txt [hunk 22]
```diff
@@ -0,0 +1,71 @@
+Meet Deepgram Aura: real-time text-to-speech for real-time AI agents
+----------
+It’s been a year since large language models (LLMs) seemingly went mainstream overnight (Happy Birthday, ChatGPT!!!), and the world has witnessed both rapid development of these technologies and immense interest in their potential. We believe that we have reached an inflection point where voice-based interfaces will be the primary means to accessing LLMs and the experiences they unlock. Here are a few recent signals in support of our thesis:
+
+- Good old fashioned voice notes are enjoying a healthy resurgence.
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/sentiment/conversation.txt:5]: this is just sample text to analyze

File: examples/prerecorded/intent/Program.cs [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+{
+    class Program
+    {
+        static async Task Main(string[] args)
+        {
+            // Initialize Library with default logging
+            // Normal logging is "Info" level
+            Library.Initialize();
+
+            // Set "DEEPGRAM_API_KEY" environment variable to your Deepgram API Key
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/prerecorded/intent/Program.cs:19]: ignoring

File: examples/prerecorded/summary/Program.cs [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+        static async Task Main(string[] args)
+        {
+            // Initialize Library with default logging
+            // Normal logging is "Info" level
+            Library.Initialize();
+
+            // Set "DEEPGRAM_API_KEY" environment variable to your Deepgram API Key
+            var deepgramClient = new PreRecordedClient();
+
+            var audioData = File.ReadAllBytes(@"CallCenterPhoneCall.mp3");
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/prerecorded/summary/Program.cs:29]: this already exists

File: examples/prerecorded/file/Program.cs [hunk 25]
```diff
@@ -22,14 +22,15 @@ static async Task Main(string[] args)
             // Set "DEEPGRAM_API_KEY" environment variable to your Deepgram API Key
             var deepgramClient = new PreRecordedClient();
 
-            var response = await deepgramClient.TranscribeUrl(
-                new UrlSource("https://static.deepgram.com/examples/Bueller-Life-moves-pretty-fast.wav"),
+            var audioData = File.ReadAllBytes(@"Bueller-Life-moves-pretty-fast.wav");
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/prerecorded/file/Program.cs:32]: this already exists. ignoring.

File: tests/edge_cases/keepalive/Program.cs [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
+{
+    class Program
+    {
+        static async Task Main(string[] args)
+        {
+            // Initialize Library with default logging
+            // Normal logging is "Info" level
+            Library.Initialize();
+            // OR very chatty logging
+            //Library.Initialize(LogLevel.Debug); // LogLevel.Default, LogLevel.Debug, LogLevel.Verbose
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply tests/edge_cases/keepalive/Program.cs:19]: intentionally left behind. this just is convenience code to switch quickly between normal logging and debug

File: examples/analyze/intent/conversation.txt [hunk 27]
```diff
@@ -0,0 +1,71 @@
+Meet Deepgram Aura: real-time text-to-speech for real-time AI agents
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/intent/conversation.txt:1]: this is just sample text to analyze. ignoring.

File: examples/prerecorded/topic/Program.cs [hunk 28]
```diff
@@ -0,0 +1,39 @@
...
+            var deepgramClient = new PreRecordedClient();
+
+            var audioData = File.ReadAllBytes(@"CallCenterPhoneCall.mp3");
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/prerecorded/topic/Program.cs:29]: this already exists

File: examples/analyze/sentiment/conversation.txt [hunk 29]
```diff
@@ -0,0 +1,71 @@
+Meet Deepgram Aura: real-time text-to-speech for real-time AI agents
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/sentiment/conversation.txt:1]: this is just sample text to analyze. ignoring.

File: examples/analyze/topic/conversation.txt [hunk 30]
```diff
@@ -0,0 +1,71 @@
...
+----------
+It’s been a year since large language models (LLMs) seemingly went mainstream overnight (Happy Birthday, ChatGPT!!!), and the world has witnessed both rapid development of these technologies and immense interest in their potential. We believe that we have reached an inflection point where voice-based interfaces will be the primary means to accessing LLMs and the experiences they unlock. Here are a few recent signals in support of our thesis:
+
+- Good old fashioned voice notes are enjoying a healthy resurgence.
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/topic/conversation.txt:5]: this is just sample text to analyze. ignoring.

File: examples/analyze/intent/conversation.txt (hunk 17 above)
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/intent/conversation.txt:5]: this is just sample text. ignoring this feedback.

File: examples/analyze/sentiment/conversation.txt (hunk 22 above)
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/sentiment/conversation.txt:5]: this is just sample text. ignoring this feedback.

File: examples/analyze/summary/conversation.txt [hunk 31]
```diff
@@ -0,0 +1,71 @@
...
+----------
+It’s been a year since large language models (LLMs) seemingly went mainstream overnight (Happy Birthday, ChatGPT!!!), and the world has witnessed both rapid development of these technologies and immense interest in their potential. We believe that we have reached an inflection point where voice-based interfaces will be the primary means to accessing LLMs and the experiences they unlock. Here are a few recent signals in support of our thesis:
+
+- Good old fashioned voice notes are enjoying a healthy resurgence.
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/summary/conversation.txt:5]: this is just sample text. ignoring this feedback.

File: examples/analyze/topic/conversation.txt (hunk 20 above)
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/topic/conversation.txt:1]: this is just sample text. ignoring this feedback.

File: examples/analyze/sentiment/conversation.txt (hunk 29 above)
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/sentiment/conversation.txt:1]: this is just sample text. ignoring this feedback.

**Review by SandraRodgers — APPROVED** [review]
